From a6860d749613c946100a79e0bb12947ff8d3839f Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:30:08 -0400
Subject: [net] openvswitch: supply a dummy err_handler of gre_cisco_protocol to prevent kernel crash

Message-id: <cc4c45f432565b816007e43690ff9705020df923.1410531783.git.jbenc@redhat.com>
Patchwork-id: 94013
O-Subject: [RHEL7.1 net PATCH 35/50] openvswitch: supply a dummy err_handler of gre_cisco_protocol to prevent kernel crash
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit e0bb8c44ed5cfcc56b571758ed966ee48779024c
Author: Wei Zhang <asuka.com@163.com>
Date:   Sat Jun 28 12:34:53 2014 -0700

    openvswitch: supply a dummy err_handler of gre_cisco_protocol to prevent kernel crash

    When use gre vport, openvswitch register a gre_cisco_protocol but
    does not supply a err_handler with it. The gre_cisco_err() in
    net/ipv4/gre_demux.c expect err_handler be provided with the
    gre_cisco_protocol implementation, and call ->err_handler() without
    existence check, cause the kernel crash.

    This patch provide a err_handler to fix this bug.
    This bug introduced by commit aa310701e787087d (openvswitch: Add gre
    tunnel support.)

    Signed-off-by: Wei Zhang <asuka.com@163.com>
    Signed-off-by: Jesse Gross <jesse@nicira.com>
    Signed-off-by: Pravin B Shelar <pshelar@nicira.com>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/openvswitch/vport-gre.c b/net/openvswitch/vport-gre.c
index 7c8c07f..dcd2bb1 100644
--- a/net/openvswitch/vport-gre.c
+++ b/net/openvswitch/vport-gre.c
@@ -111,6 +111,22 @@ static int gre_rcv(struct sk_buff *skb,
  return PACKET_RCVD;
 }
 
+/* Called with rcu_read_lock and BH disabled. */
+static int gre_err(struct sk_buff *skb, u32 info,
+     const struct tnl_ptk_info *tpi)
+{
+ struct ovs_net *ovs_net;
+ struct vport *vport;
+
+ ovs_net = net_generic(dev_net(skb->dev), ovs_net_id);
+ vport = rcu_dereference(ovs_net->vport_net.gre_vport);
+
+ if (unlikely(!vport))
+  return PACKET_REJECT;
+ else
+  return PACKET_RCVD;
+}
+
 static int gre_tnl_send(struct vport *vport, struct sk_buff *skb)
 {
  struct net *net = ovs_dp_get_net(vport->dp);
@@ -187,6 +203,7 @@ error:
 
 static struct gre_cisco_protocol gre_protocol = {
  .handler        = gre_rcv,
+ .err_handler    = gre_err,
  .priority       = 1,
 };
 
-- 
1.7.1